### PR TITLE
update dmodule to use slices rather than cstrings where possible

### PR DESCRIPTION
**Pull request by benjones**
It looks like Module.create is part of the C++ API so can't take a slice?  Should it be marked extern(C++) ?

### CODE REVIEW

**Comment by thewilsonator** [comment]
> It looks like Module.create is part of the C++ API 

That was what the create functions were for.

> so can't take a slice? 

Yes.

> Should it be marked extern(C++) ?

Is it not? This would imply that it is unused by LDC & GDC. If that is the case it can probably be removed.

**Review by Geod24 — DISMISSED** [review]
Need to update the C++ header

**Comment by jacob-carlborg** [comment]
> It looks like Module.create is part of the C++ API so can't take a slice?

No. But it would be better to have one overload taking a slice and one taking string and length separately which will forward to the overload taking a slice. Or wrap the pointer and length in a struct.

**Comment by benjones** [comment]
@thewilsonator I was asking if it should be marked at the definition site, but the class is marked as extern C++, so nevermind

@Geod24 It looks like I just need to update the declaration for arg in module.h, but how do you declare a slice member in C++?  Am I missing something else?

**Comment by jacob-carlborg** [comment]
> It looks like I just need to update the declaration for arg in module.h, but how do you declare a slice member in C++? Am I missing something else?

There's should be a struct somewhere, perhaps array.h, that you can use for a slice.

File: src/dmd/dmodule.d [hunk 1]
```diff
@@ -405,39 +405,51 @@ extern (C++) final class Module : Package
...
+            FileName.free(srcfilename.ptr);
             srcfilename = FileName.removeExt(filename); // just does a mem.strdup(filename)
         }
-        else if (!FileName.equalsExt(srcfilename, global.mars_ext) && !FileName.equalsExt(srcfilename, global.hdr_ext) && !FileName.equalsExt(srcfilename, "dd"))
+        else if (!FileName.equalsExt(srcfilename, global.mars_ext.toDString()) &&
+                 !FileName.equalsExt(srcfilename, global.hdr_ext.toDString()) &&
+                 !FileName.equalsExt(srcfilename, "dd"))
         {
-            error("source file name '%s' must have .%s extension", srcfilename, global.mars_ext);
+            error("source file name '%s' must have .%s extension", srcfilename.ptr, global.mars_ext);
```
Comment (WalterBright) [inline src/dmd/dmodule.d]: This relies on .ptr being 0 terminated. But I don't see where that guarantee is made.

File: src/dmd/dmodule.d [hunk 2]
```diff
@@ -405,39 +405,51 @@ extern (C++) final class Module : Package
...
+    extern (D) this(const ref Loc loc, const(char)[] filename, Identifier ident, int doDocComment, int doHdrGen)
     {
         super(loc, ident);
-        const(char)* srcfilename;
+        const(char)[] srcfilename;
         //printf("Module::Module(filename = '%s', ident = '%s')\n", filename, ident.toChars());
         this.arg = filename;
-        srcfilename = FileName.defaultExt(filename, global.mars_ext);
-        if (global.run_noext && global.params.run && !FileName.ext(filename) && FileName.exists(srcfilename) == 0 && FileName.exists(filename) == 1)
+        srcfilename = FileName.defaultExt(filename, global.mars_ext.toDString());
```
Comment (WalterBright) [inline src/dmd/dmodule.d]: I see all these `toDString()` calls everywhere, which do a `strlen()`. Not sure this is an improvement. Also, `mars_ext` should be declared as a string, and avoid the `toDString()`s.

**Review by WalterBright — DISMISSED** [review]
see comments

**Comment by benjones** [comment]
@WalterBright  The FileName class and its methods seem to guarantee slices should all be null terminated, so passing .ptr as a CString should be safe.  I'll double check though.  At a first glance, the uses of .ptr in glue.d scare me because they come from arg which did not come from a FileName method.

As for the calls toDString, a lot of the overloads that take cstrings immediately call toDString on all the arguments anyway.  For example, the setOutFile calls have reduced the net number of toDString calls by 1.

I was going to look at stuff like mars_ext next, but didn't want to put it all in one PR.  From here, grepping for .toDString and .ptr should give a list of smells to fix up.

File: src/dmd/dmodule.d (hunk 2 above)
Comment (jacob-carlborg) [reply src/dmd/dmodule.d]: Until all C strings have been changed to D strings I think we have to live with `toDString` and some inefficiencies like calling `strlen`.

File: src/dmd/dmodule.d (hunk 1 above)
Comment (benjones) [reply src/dmd/dmodule.d]: These strings come from FileName and its methods, which make that guarantee.  I'll continue to try to switch stuff to slices so we can eliminate spamming .ptr everywhere.

File: src/dmd/dmodule.d (hunk 2 above)
Comment (benjones) [reply src/dmd/dmodule.d]: As I mentioned in another comment, defaultExt and SetOutFile immediately call toDstring on everything for the char* overload, so we're actually decreasing the net strlen calls here.  I'll continue to slicify things to remove those.

**Comment by benjones** [comment]
I think I addressed @WalterBright 's and @Geod24 's comments.  I plan on reducing the calls to .ptr and .toDString in subsequent PRs unless people feel strongly that I should do so here.

**Comment by thewilsonator** [comment]
This will have to wait until (maybe after) dconf because the other reviewers are travelling ATM. Apologies for the wait. Ping me if I happen to forget.

File: src/dmd/dmodule.d (hunk 1 above)
Comment (jacob-carlborg) [reply src/dmd/dmodule.d]: Please pass the length and pointer separately to `printf`-like functions. Then 0 termination doesn’t matter.

**Comment by benjones** [comment]
I think I removed reliance on null terminated strings everywhere.  I'm working on a PR to constants in global.d DStrings which should eliminate .toDString noise

**Review by jacob-carlborg — APPROVED** [review]

**Pull request merged**